src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs

```diff
@@ -1,93 +1,213 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System.Collections.Generic;
 using System.Composition;
 using System.Linq;
-using System.Threading;
 using Microsoft.CodeAnalysis.CodeRefactorings;
 using Microsoft.CodeAnalysis.CodeRefactorings.InvertIf;
 using Microsoft.CodeAnalysis.CSharp.Extensions;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
-using Microsoft.CodeAnalysis.Editing;
-using Microsoft.CodeAnalysis.Formatting;
-using Microsoft.CodeAnalysis.LanguageServices;
-using Microsoft.CodeAnalysis.Operations;
-using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.InvertIf
 {
     [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.InvertIf), Shared]
-    internal partial class CSharpInvertIfCodeRefactoringProvider : AbstractInvertIfCodeRefactoringProvider
+    internal sealed class CSharpInvertIfCodeRefactoringProvider : AbstractInvertIfCodeRefactoringProvider<IfStatementSyntax, StatementSyntax>
     {
-        protected override SyntaxNode GetIfStatement(TextSpan textSpan, SyntaxToken token, CancellationToken cancellationToken)
+        protected override string GetTitle()
+            => CSharpFeaturesResources.Invert_if;
+
+        protected override bool IsElseless(IfStatementSyntax ifNode)
+            => ifNode.Else == null;
+
+        protected override bool CanInvert(IfStatementSyntax ifNode)
+            => ifNode.IsParentKind(SyntaxKind.Block, SyntaxKind.SwitchSection);
+
+        protected override SyntaxNode GetCondition(IfStatementSyntax ifNode)
+            => ifNode.Condition;
+
+        protected override StatementRange GetIfBodyStatementRange(IfStatementSyntax ifNode)
+            => new StatementRange(ifNode.Statement, ifNode.Statement);
+
+        protected override bool IsStatementContainer(SyntaxNode node)
+            => node.IsKind(SyntaxKind.Block, SyntaxKind.SwitchSection);
+
+        protected override bool IsNoOpSyntaxNode(SyntaxNode node)
+            => node.IsKind(SyntaxKind.Block, SyntaxKind.EmptyStatement);
+
+        protected override bool IsStatement(SyntaxNode node)
+            => node is StatementSyntax;
+
+        protected override SyntaxNode GetNextExecutableStatement(SyntaxNode node)
```

Comment:
is you have syntax facts service, you can get ISyntaxFactsService and ask GetNextexecutableStatement? rather than have override and 2 language implements this? is VB behave differently?

Reply:
Yes, ISyntaxFactsService.GetNextExecutableStatement did not behave as I expected in VB (sometimes returns the parent statement e.g. for the last statement in `Case` clause)

Reply:
I would not put this stuff in ISyntaxFacts.